diff -Naur cog-jglobus-1.8.0.orig/src/org/globus/gsi/CertificateRevocationLists.java cog-jglobus-1.8.0/src/org/globus/gsi/CertificateRevocationLists.java
--- cog-jglobus-1.8.0.orig/src/org/globus/gsi/CertificateRevocationLists.java	2009-11-18 16:32:50.000000000 +0100
+++ cog-jglobus-1.8.0/src/org/globus/gsi/CertificateRevocationLists.java	2010-10-14 10:21:02.219406003 +0200
@@ -237,8 +237,8 @@
              long now = System.currentTimeMillis();
              if (lastRefresh + lifetime <= now) {
 		 reload(getDefaultCRLLocations());
+                 lastRefresh = now;
 	     }
-	     lastRefresh = now;
         }
 
         private static synchronized String getDefaultCRLLocations() {
