// polymer-build/lib/polymer-build.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var analyzer_1 = require("./analyzer");
exports.BuildAnalyzer = analyzer_1.BuildAnalyzer;
var babel_plugin_bare_specifiers_1 = require("./babel-plugin-bare-specifiers");
exports.resolveBareSpecifiers = babel_plugin_bare_specifiers_1.resolveBareSpecifiers;
var bundle_1 = require("./bundle");
exports.BuildBundler = bundle_1.BuildBundler;
var custom_elements_es5_adapter_1 = require("./custom-elements-es5-adapter");
exports.addCustomElementsEs5Adapter = custom_elements_es5_adapter_1.addCustomElementsEs5Adapter;
var fork_stream_1 = require("./fork-stream");
exports.forkStream = fork_stream_1.forkStream;
var html_splitter_1 = require("./html-splitter");
exports.HtmlSplitter = html_splitter_1.HtmlSplitter;
var html_transform_1 = require("./html-transform");
exports.htmlTransform = html_transform_1.htmlTransform;
var js_transform_1 = require("./js-transform");
exports.jsTransform = js_transform_1.jsTransform;
var optimize_streams_1 = require("./optimize-streams");
exports.getOptimizeStreams = optimize_streams_1.getOptimizeStreams;
var polymer_project_1 = require("./polymer-project");
exports.PolymerProject = polymer_project_1.PolymerProject;
var service_worker_1 = require("./service-worker");
exports.addServiceWorker = service_worker_1.addServiceWorker;
exports.generateServiceWorker = service_worker_1.generateServiceWorker;
