diff -Naur a/jglobus/src/org/globus/gsi/proxy/ProxyPathValidator.java b/jglobus/src/org/globus/gsi/proxy/ProxyPathValidator.java
--- a/jglobus/src/org/globus/gsi/proxy/ProxyPathValidator.java	2008-07-09 19:00:46.000000000 +0200
+++ b/jglobus/src/org/globus/gsi/proxy/ProxyPathValidator.java	2009-12-10 10:06:38.958967887 +0100
@@ -901,11 +901,20 @@
 	//check date validity of CRL
         boolean validCRL = checkCRLValidity(crl);
         if (validCRL) {
-	    if (crl.isRevoked(cert)) {
-		throw new 
-                    ProxyPathValidatorException(ProxyPathValidatorException
-                                                .REVOKED, cert,
-                                                i18n.getMessage("proxyErr21",
-                                                                cert.getSubjectDN().getName()));
+            /* One would have thought that a CRL is immutable and thus
+             * thread safe, however inside the ASN1 parse tree we find
+             * LazyDERSequence. LazyDERSequence is parsed lazily and
+             * does so in a non-thread safe manner. One may very well
+             * classify this as a bouncy castle bug, but as a
+             * workaround synchronizing on the CRL solves the problem.
+             */
+            synchronized (crl) {
+                if (crl.isRevoked(cert)) {
+                    throw new
+                        ProxyPathValidatorException(ProxyPathValidatorException
+                                                    .REVOKED, cert,
+                                                    i18n.getMessage("proxyErr21",
+                                                                    cert.getSubjectDN().getName()));
+                }
             }
 	} else {
