diff -Naur jglobus.old/src/org/globus/gsi/gssapi/GlobusGSSContextImpl.java jglobus/src/org/globus/gsi/gssapi/GlobusGSSContextImpl.java
--- jglobus.old/src/org/globus/gsi/gssapi/GlobusGSSContextImpl.java	2008-01-11 00:12:55.000000000 +0100
+++ jglobus/src/org/globus/gsi/gssapi/GlobusGSSContextImpl.java	2009-06-03 12:58:07.973127209 +0200
@@ -752,13 +752,7 @@
                                          "noCaCerts");
         }
         try {
-            if (this.tc instanceof PureTLSTrustedCertificates) {
-                this.context.setRootList( 
-                    ((PureTLSTrustedCertificates)this.tc).getX509CertList()
-                );
-            } else {
-                this.context.setTrustedCertificates(this.tc.getCertificates());
-            }
+            this.context.setRootList(this.tc.getX509CertList());
         } catch (GeneralSecurityException e) {
             throw new GlobusGSSException(GSSException.FAILURE, e);
         }
diff -Naur jglobus.old/src/org/globus/gsi/TrustedCertificates.java jglobus/src/org/globus/gsi/TrustedCertificates.java
--- jglobus.old/src/org/globus/gsi/TrustedCertificates.java	2008-04-09 16:57:23.000000000 +0200
+++ jglobus/src/org/globus/gsi/TrustedCertificates.java	2009-06-03 12:58:07.973127209 +0200
@@ -15,6 +15,7 @@
  */
 package org.globus.gsi;
 
+import java.security.GeneralSecurityException;
 import java.security.cert.X509Certificate;
 import java.util.Map;
 import java.util.Vector;
@@ -22,12 +23,14 @@
 import java.util.StringTokenizer;
 import java.util.Collection;
 import java.util.Iterator;
+import java.util.Vector;
 import java.io.File;
 import java.io.IOException;
 import java.io.FilenameFilter;
 
 import org.globus.common.CoGProperties;
 import org.globus.util.TimestampEntry;
+import org.globus.gsi.ptls.PureTLSUtil;
 
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
@@ -51,6 +54,7 @@
     private Map certSubjectDNMap;
     private Map certFileMap;
     private boolean changed;
+    private Vector certList = null;
 
     // DN is in Globus format here, without any reversal.
     private Map policyDNMap;
@@ -90,6 +94,19 @@
         }
     }
 
+    /**
+     * Returns the trusted certificates as a Vector of X509Cert objects.
+     */
+    public synchronized Vector getX509CertList()
+        throws GeneralSecurityException
+    {
+      if (certList == null) {
+            certList =
+                PureTLSUtil.certificateChainToVector(getCertificates());
+      }
+      return certList;
+    }
+
     public X509Certificate[] getCertificates() {
         if (this.certSubjectDNMap == null) {
             return null;
@@ -274,6 +291,7 @@
 
         this.certFileMap = newCertFileMap;
         this.certSubjectDNMap = newCertSubjectDNMap;
+        this.certList = null;
 
         this.policyFileMap = newSigningFileMap;
         this.policyDNMap = newSigningDNMap;
@@ -293,6 +311,7 @@
                           Map newPolicyFileMap, Map newPolicyDNMap) {
 
         X509Certificate cert = null;
+        this.certList = null;
         
         if (this.certFileMap == null) {
             this.certFileMap = new HashMap();
