--- dcap-2.47.5.orig/src/dcap_reconnect.c	2010-09-20 16:28:23.971844000 +0200
+++ dcap-2.47.5/src/dcap_reconnect.c	2010-11-06 19:25:14.733554183 +0100
@@ -74,7 +74,7 @@
 	}
 	m_unlock(&kLock);
 
-	t_getspecific(ioFailedKey, (void **)&io);
+	t_getspecific(ioFailedKey, &io);
 	if( io == NULL ) {
 		io = calloc(1, sizeof(int));
 		t_setspecific(ioFailedKey, (void *)io);
@@ -96,7 +96,7 @@
 	}
 	m_unlock(&kLock);
 
-	t_getspecific(isAlarmKey, (void **)&al);
+	t_getspecific(isAlarmKey, &al);
 	if( al == NULL ) {
 		al = calloc(1, sizeof(int));
 		t_setspecific(isAlarmKey, (void *)al);
@@ -118,7 +118,7 @@
 	}
 	m_unlock(&kLock);
 
-	t_getspecific(sa_alarmKey, (void **)&sa);
+	t_getspecific(sa_alarmKey, &sa);
 	if( sa == NULL ) {
 		sa = calloc(1, sizeof(struct sigaction));
 		t_setspecific(sa_alarmKey, (void *)sa);
